File: Data Structures/Binary Search Tree/Implementation.go
// Declare the main package
package main

// Define a struct for the nodes in the binary search tree
type TreeNode struct {
	Val   int       // The value stored in the node
	Left  *TreeNode // Pointer to the left child node
	Right *TreeNode // Pointer to the right child node
}

// Function to create a new tree node
func NewTreeNode(val int) *TreeNode {
	return &TreeNode{
		Val:   val, // Set the node's value
		Left:  nil, // Initialize the left child as nil
		Right: nil, // Initialize the right child as nil
	}
}

// Define a struct for the search operation
type Search struct{}

// Method to perform a binary search in the tree
func (s *Search) Search(root *TreeNode, target int) bool {
	if root == nil { // If the tree or subtree is empty, the target is not found
		return false
	}

	// Compare the target with the root's value
	if target > root.Val { // If the target is greater, search the right subtree
		return s.Search(root.Right, target)
	} else if target < root.Val { // If the target is smaller, search the left subtree
		return s.Search(root.Left, target)
	} else { // If the target is equal to the root's value, the target is found
		return true
	}
}

// Time complexity: O(h), where h is the height of the tree
// Space complexity: O(h), where h is the height of the tree

File: Data Structures/Binary Search Tree/Implementation.py
# Define a class for the nodes in the binary search tree
class TreeNode:
    # Initialize a tree node with a value and two children
    def __init__(self, val):
        self.val = val  # The value stored in the node
        self.left = None  # Initialize the left child as None
        self.right = None  # Initialize the right child as None

# Function to perform a binary search in the tree
def search(root, target):
    if not root:  # If the tree or subtree is empty, the target is not found
        return False

    # Compare the target with the root's value
    if target > root.val:  # If the target is greater, search the right subtree
        return search(root.right, target)
    elif target < root.val:  # If the target is smaller, search the left subtree
        return search(root.left, target)
    else:  # If the target is equal to the root's value, the target is found
        return True
    
# time complexity: O(h), where h is the height of the tree
# space complexity: O(h), where h is the height of the tree